docs/sql_reference/src/asciidoc/_chapters/sql_functions_and_expressions.adoc

```diff
@@ -487,6 +487,141 @@ LARGEINT if the precision of the argument is greater than or equal to
 ABS (-20 + 12)
 ```
 
+<<<
+[[aes_encrypt_function]]
+== AES_ENCRYPT Function
+ 
+The AES_ENCRYPT function encrypts a `_string_` using a specified encryption `_key_` with the AES (Advanced Encryption Standard) algorithm and returns a binary string. 
+
+This function prevents sensitive data from being visible to the public. For example, use the AES_ENCRYPT function to encrypt person identity number to preserve user confidentiality.
+
+`AES_ENCRYPT (_string_, _key_)`
+
+[[syntax_description_of_aes_encrypt]]
+=== Syntax Description of AES_ENCRYPT
+
+* `_string_` 
++
+is the data to be encrypted. 
+
+* `_key_` 
++
+is the encryption key to encrypt the `_string_`.
+
++
+TIP: The CQD `block_encryption_mode` controls the mode for the block-based encryption algorithm. The default mode is the aes-128-ecb, which means the encryption using a key length of 128 bits and the ECB mode.
+
++
+[cols="50%,50%"]
+|===
+^| *Value* ^| *Mode*
+^| 0 ^| aes-128-ecb
+^| 1 ^| aes-192-ecb
+^| 2 ^| aes-256-ecb
+|===
+
++
+*Example*
+
++
+The following examples show that the AES_ENCRYPT function returns the different results based on the different modes. 
+
++
+** The mode aes_192_ecb is in effect.
+
++
+```
+CQD BLOCK_ENCRYPTION_MODE '1';
+
+--- SQL operation complete.
+```
+
++
+```
+SELECT AES_ENCRYPT ('Technical_Writer','Coder') FROM DUAL;
+
+(EXPR)
+--------------------------------
+4}\îî¢Ø­ã^ûh<á¼Æ7õé³ê!cH^Ð
+```
+
++
+** The mode aes_256_ecb is in effect. 
+
++
+```
+CQD BLOCK_ENCRYPTION_MODE '2';
+
+--- SQL operation complete.
+```
+
++
+```
+SELECT AES_ENCRYPT ('Technical_Writer','Coder') FROM DUAL;
+
+(EXPR)
+--------------------------------
+Ô_8ÆðÑ=Uú2g1¥@°Ô3yä­
+
+--- 1 row(s) selected.
+```
+
+[[considerations_for_aes_encrypt]]
+=== Considerations for AES_ENCRYPT 
+
+Either argument cannot be NULL. 
```

Comment:
Perhaps: "Neither argument can be NULL."
Are there limits to how long string can be?  Any restrictions on what can be contained in str?
Any rules on what can be used as a key? 
Do we know the length of the encrypted value?  Is it less than or equal to the string?

Reply:
Thanks Roberta, I've incorporated your comments.

The developer for the AES_ENCRYPT function and the AES_DECRYPT function is a ghost user in GitHub, so I've tested them by myself.

Q1. Are there limits to how long string can be?
I don't know how to test the accurate length for the `string`, I've tried 0, 1 ... 100, and they all work.
MySQL Manual says the `string` and the `key` can be any length, I don't know whether it is appropriate for Trafodion.

Q2. Any restrictions on what can be contained in str? 
Q3. Any rules on what can be used as a key?
I've tested and found that the `string` and the `key` are case-sensitive and can contain a mix of uppercase and lowercase letters, and includes numbers and special characters. I've added these descriptions.

Q4. Do we know the length of the encrypted value? 
Q5. Is it less than or equal to the string?
MySQL Manual says:
> The str and crypt_str arguments can be any length, and padding is automatically added to str
so it is a multiple of a block as required by block-based algorithms such as AES. This padding is
automatically removed by the AES_DECRYPT() function. The length of crypt_str can be calculated
using this formula: 16 * (trunc(string_length/16) + 1).

I've tested it on Trafodion as below, it seems that this formula is not appropriate for Trafodion, I don't know how to get the correct formula, could you please help?

- The length of the `string` _aaaaaaaaaaaaaaaa_ is 16, and the length of the result (encrypted value) is 24.
```
SELECT AES_ENCRYPT ('aaaaaaaaaaaaaaaa','b') FROM DUAL;

(EXPR)
--------------------------------
P+¯{ÄPCöúúCüü|ñXWV(ÿÿçIÁ

--- 1 row(s) selected.
```

- The length of the `string` _aaaaaaaaaaaaaaaa_ is 16, and the length of the result (encrypted value) is 21.
```
SELECT AES_ENCRYPT ('aaaaaaaaaaaaaaaa','8') FROM DUAL;

(EXPR)
--------------------------------
0ÃbA>óWý°à3"?!ò!êyÞÅÜ

--- 1 row(s) selected.
```
- The length of the `string` _aaaaaaaaaaaaaaaa_ is 16, and the length of the result (encrypted value) is 20.
```
SELECT AES_ENCRYPT ('aaaaaaaaaaaaaaaa','bB123&*') FROM DUAL;

(EXPR)
--------------------------------
(-;·ÚàÍ<+±PLO½r;è%{Í 

--- 1 row(s) selected.
```

Reply:
It looks like the function was added by Weiqing Xu, not sure who is familiar with this function now.  I will check around. 

Here is the message from his delivery, it looks like there are more values for the CQD:

   [TRAFODION-2228]Add AES_ENCRYPT/AES_DECRYPT functions
    
    Add two functions which used to encrypt and decrypt the data using the
      official AES (Advanced Encryption Standard) algorithm.
    AES_ENCRYPT(str, key_str[, init_vector])
    AES_DECRYPT(crypt_str, key_str[, init_vector])
    
    Both these two function have optional init_vector argument which
    provides an initalization vector for block encryption modes that
    require it. For modes that require the optional init_vector argument, an err
    occurs if init_vector is missing. For modes that don't need
    init_vector, it will raise a unused argument waring if init_vector is
    provided.
    
    The new CQD block_encryption_mode controls the mode for
    block-based
    encryption algorithm. The default value is 0, which use
    aes-128-ecb.
   
    CQD value          algorithm
    0             aes-128-ecb
    1             aes_192_ecb
    2             aes_256_ecb
    3             aes_128_cbc
    4             aes_192_cbc
    5             aes_256_cbc
    6             aes_128_cfb1
    7             aes_192_cfb1
    8             aes_256_cfb1
    9             aes_128_cfb8
    10            aes_192_cfb8
    11            aes_256_cfb8
    12            aes_128_cfb128
    13            aes_192_cfb128
    14            aes_256_cfb128
    15            aes_128_ofb
    16            aes_192_ofb
    17            aes_256_ofb

Reply:
It looks like openssl defines the max key length to use.  Its size depends on the version of openSSL installed.  On development systems it is 64 (/usr/bin/openssl/evp.h) but it looks like it can go up to 512 and beyond (https://github.com/openssl/openssl/issues/4777).

I looked up openssl here is what it states for encrypted length (https://www.openssl.org/docs/man1.0.2/man3/EVP_EncryptUpdate.html):

EVP_EncryptUpdate() encrypts inl bytes from the buffer in and writes the encrypted version to out. This function can be called multiple times to encrypt successive blocks of data. The amount of data written depends on the block alignment of the encrypted data: as a result the amount of data written may be anything from zero bytes to (inl + cipher_block_size - 1) so out should contain sufficient room. The actual number of bytes written is placed in outl.

So, I guess for now, not mention the encrypted length until we can come up with a formula.

Reply:
1. CQD values.
Yes, the CQD contains 17 values as I mentioned in the first comment of this PR:
> The PR #874 shows that the CQD block_encryption_mode has 17 values and corresponding modes. I've tested each of them on Trafodion 2.4, only 3 modes (aes-128-ecb, aes_192_ecb, and aes_256_ecb) are effective, so I've documented these.

2. Key length.
OK, thank you for your guidance.